### PR TITLE
Added alpha matting sample

### PR DESCRIPTION
**Pull request by HarxSan**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: samples/dnn/alpha_matting.cpp [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+    int input_width = parser.has("width") ? parser.get<int>("width") : 512;
+    int input_height = parser.has("height") ? parser.get<int>("height") : 512;
+    float scale_factor = parser.has("scale") ? parser.get<float>("scale") : 1.0f;
+    Scalar mean_values = parser.has("mean") ? parser.get<Scalar>("mean") : Scalar(0.0, 0.0, 0.0);
+    bool swapRB = parser.has("rgb") ? parser.get<bool>("rgb") : true;
+    String backend = parser.get<String>("backend");
+    String target = parser.get<String>("target");
+    String sha1 = parser.has("sha1") ? parser.get<String>("sha1") : "";
+
+    string model = findModel(parser.get<String>("model"), sha1);
```
Comment (abhishek-gola) [inline samples/dnn/alpha_matting.cpp]: Add these parameters to models.yml and read from that.

File: samples/dnn/alpha_matting.cpp [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+ * This sample demonstrates human portrait alpha matting using MODNet model.
+ * MODNet is a trimap-free portrait matting method that can produce high-quality
+ * alpha mattes for portrait images in real-time.
+ *
+ * Usage:
+ *   ./alpha_matting --input=image.jpg                      # Process image
+ *   ./alpha_matting --input=video.mp4                      # Process video
+ *   ./alpha_matting                                        # Use webcam
+ *   ./alpha_matting modnet                                 # Use config alias
+ *
```
Comment (abhishek-gola) [inline samples/dnn/alpha_matting.cpp]: ./example_dnn_alpha_matting instead of ./alpha_matting

File: samples/dnn/alpha_matting.cpp [hunk 3]
```diff
@@ -0,0 +1,247 @@
+/**
+ * @file alpha_matting.cpp
```
Comment (abhishek-gola) [inline samples/dnn/alpha_matting.cpp]: On top, add opencv license.

```
// This file is part of OpenCV project.
// It is subject to the license terms in the LICENSE file found in the top-level directory
// of this distribution and at http://opencv.org/license.html.
```

File: samples/dnn/alpha_matting.cpp [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+        "\t./alpha_matting --input=image.jpg\n"
+        "\t./alpha_matting --input=video.mp4\n"
+        "\t./alpha_matting (for webcam)\n"
+        "\t./alpha_matting modnet (using config alias)\n\n"
+        "Download MODNet model from: https://github.com/ZHKKKe/MODNet/releases\n"
+        "Press 'q' or ESC to quit \n";
+
+    const string param_keys =
+        "{ help h          |                   | Print help message }"
+        "{ @alias          |                   | An alias name of model to extract preprocessing parameters from models.yml file }"
```
Comment (abhishek-gola) [inline samples/dnn/alpha_matting.cpp]: default alias can be added here.

File: samples/dnn/alpha_matting.cpp [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+
+    const char *path = getenv("OPENCV_SAMPLES_DATA_PATH");
+    if ((path != NULL) || parser.has("@alias") || (parser.get<String>("model") != ""))
+    {
+        if (modelName.empty())
+        {
+            modelName = "modnet"; // Default alias
+        }
+        zooFile = findFile(zooFile);
+    }
```
Comment (abhishek-gola) [inline samples/dnn/alpha_matting.cpp]: As alias is being defined above, no need of this.

File: samples/dnn/alpha_matting.cpp [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+        if (key == 'q' || key == 27)
+        {
+            break;
+        }
+    }
+
+    cap.release();
+    destroyAllWindows();
+    return 0;
+}
```
Comment (abhishek-gola) [inline samples/dnn/alpha_matting.cpp]: Add an empty line in the end of file.

File: samples/dnn/alpha_matting.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    return net
+
+def apply_modnet(model, image):
+    out = model.forward()
+    alpha_mask, composite = postprocess_output(image, out)
+    t, _ = model.getPerfProfile()
+    label = 'Inference time: %.2f ms' % (t * 1000.0 / cv.getTickFrequency())
+    cv.putText(image, label, (0, 15), cv.FONT_HERSHEY_SIMPLEX, 0.5, (0, 255, 0))
+    cv.putText(alpha_mask, label, (0, 15), cv.FONT_HERSHEY_SIMPLEX, 0.5, (255, 255, 255))
+    cv.putText(composite, label, (0, 15), cv.FONT_HERSHEY_SIMPLEX, 0.5, (0, 255, 0))
```
Comment (abhishek-gola) [inline samples/dnn/alpha_matting.py]: Make it dynamic to image size.

File: samples/dnn/alpha_matting.cpp [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+    if (!cap.isOpened())
+    {
+        cerr << "Error: Cannot open input source" << endl;
+        return -1;
+    }
+
+    setupWindows();
+
+    cout << "Press 'q' or ESC to quit, 's' to save current frame results" << endl;
+
```
Comment (abhishek-gola) [inline samples/dnn/alpha_matting.cpp]: putText it to image/frame as well.

File: samples/dnn/alpha_matting.py [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+    if not cap.isOpened():
+        print("Failed to open the input video")
+        exit(-1)
+    cv.namedWindow('Input', cv.WINDOW_AUTOSIZE)
+    cv.namedWindow('Alpha Mask', cv.WINDOW_AUTOSIZE)
+    cv.namedWindow('Composite', cv.WINDOW_AUTOSIZE)
+    cv.moveWindow('Alpha Mask', 200, 50)
+    cv.moveWindow('Composite', 400, 50)
+
+    net = None
```
Comment (abhishek-gola) [inline samples/dnn/alpha_matting.py]: Not a good idea to initialise net as None, directly read the model and assign to net.

**Comment by abhishek-gola** [comment]
Add models.yml to your pull request.

File: samples/dnn/alpha_matting.cpp (hunk 1 above)
Comment (HarxSan) [reply samples/dnn/alpha_matting.cpp]: Updated to read those parameters from models.yml

File: samples/dnn/alpha_matting.cpp (hunk 2 above)
Comment (HarxSan) [reply samples/dnn/alpha_matting.cpp]: Changes has been made.

File: samples/dnn/alpha_matting.cpp (hunk 3 above)
Comment (HarxSan) [reply samples/dnn/alpha_matting.cpp]: Included opencv license.

File: samples/dnn/alpha_matting.cpp (hunk 4 above)
Comment (HarxSan) [reply samples/dnn/alpha_matting.cpp]: default alias set to modnet.

File: samples/dnn/alpha_matting.cpp (hunk 5 above)
Comment (HarxSan) [reply samples/dnn/alpha_matting.cpp]: Removed unnecessary alias.

File: samples/dnn/alpha_matting.cpp (hunk 6 above)
Comment (HarxSan) [reply samples/dnn/alpha_matting.cpp]: Added a new line.

File: samples/dnn/alpha_matting.cpp [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+ * Usage:
+ *   ./example_dnn_alpha_matting --input=image.jpg                      # Process image
+ *   ./example_dnn_alpha_matting --input=video.mp4                      # Process video
+ *   ./example_dnn_alpha_matting                                        # Use webcam
+ *   ./example_dnn_alpha_matting modnet                                 # Use config alias
+ *
+ * Requirements:
+ *   - OpenCV >= 4.5.0 with DNN module
+ *   - MODNet ONNX model
+ */
```
Comment (abhishek-gola) [inline samples/dnn/alpha_matting.cpp]: Have tested it on 4.x branch? because some models only work for 5.x branch, as you are pushing it to 5.x branch I just wanted to confirm if it works on 4.x if not then update the requirement to >= 5.0.0

File: samples/dnn/alpha_matting.py (hunk 7 above)
Comment (HarxSan) [reply samples/dnn/alpha_matting.py]: Made it dynamic based on the image size.

File: samples/dnn/alpha_matting.py [hunk 11]
```diff
@@ -0,0 +1,146 @@
+"""
```
Comment (abhishek-gola) [inline samples/dnn/alpha_matting.py:1]: Add license here too, along with this bigvision copyright.

```
This file is part of OpenCV project.
It is subject to the license terms in the LICENSE file found in the top-level directory
of this distribution and at http://opencv.org/license.html.

Copyright (C) 2025, Bigvision LLC.
```

File: samples/dnn/alpha_matting.cpp (hunk 8 above)
Comment (HarxSan) [reply samples/dnn/alpha_matting.cpp]: Text is now displayed in the input image/frame.

File: samples/dnn/alpha_matting.py (hunk 9 above)
Comment (HarxSan) [reply samples/dnn/alpha_matting.py]: I've removed the net = None line and now assign the loaded model directly to net.

**Comment by HarxSan** [comment]
> Add models.yml to your pull request.

Added models.yml to the pull request.

File: samples/dnn/alpha_matting.cpp (hunk 10 above)
Comment (HarxSan) [reply samples/dnn/alpha_matting.cpp]: Haven't tested yet on 4.x branch, will update the requirement to >= 5.0.0

File: samples/dnn/alpha_matting.py (hunk 11 above)
Comment (HarxSan) [reply samples/dnn/alpha_matting.py:1]: Added license along with bigvision copyright.

**Comment by abhishek-gola** [comment]
Hi @HarxSan,
Good job! These changes are looking good, I will just test them on my local and let you know if they are working as desired or not.

**Comment by HarxSan** [comment]
> Hi @HarxSan, Good job! These changes are looking good, I will just test them on my local and let you know if they are working as desired or not.

Thanks @abhishek-gola ! Let me know if you face any issues during testing.

File: samples/dnn/alpha_matting.cpp [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+
+
+    string model = findModel(parser.get<String>("model"), sha1);
+
+    if (model.empty())
+    {
+        cout << "[ERROR] MODNet model not found!" << endl;
+        cout << "Please download MODNet ONNX model from: https://github.com/ZHKKKe/MODNet/releases" << endl;
+        cout << "Or specify model path using --model argument" << endl;
+        return -1;
```
Comment (abhishek-gola) [inline samples/dnn/alpha_matting.cpp]: Mention in description to firstly download model using `python download_models.py modnet`. And if model is not found _findModel_ will automatically throw the error, hence no need of this code block.

File: samples/dnn/alpha_matting.cpp [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+        processFrame(frame, alpha_mask, composite, net, scale_factor, input_width, input_height, mean_values, swapRB);
+
+        string helpText = "Press 'q' or ESC to quit";
+        int fontFace = cv::FONT_HERSHEY_SIMPLEX;
+        double fontScale = std::max(frame.cols, frame.rows) / 1000.0;
+        int thickness = 1;
+        int baseline = 0;
+        cv::Size textSize = cv::getTextSize(helpText, fontFace, fontScale, thickness, &baseline);
+        cv::Point textOrg(10, textSize.height + 10);
+        cv::putText(frame, helpText, textOrg, fontFace, fontScale, cv::Scalar(0, 255, 0), thickness);
```
Comment (abhishek-gola) [inline samples/dnn/alpha_matting.cpp]: If it's just about _`quit`_ then no need to putText it.

File: samples/dnn/alpha_matting.cpp [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+    cout << "Press 'q' or ESC to quit" << endl;
+
+    Mat frame, alpha_mask, composite;
+
+    for (;;)
+    {
+        cap >> frame;
+        if (frame.empty())
+        {
+            cout << "End of video or cannot read frame" << endl;
```
Comment (abhishek-gola) [inline samples/dnn/alpha_matting.cpp]: Add `waitKey(0);` here else it will get terminate quickly in case of image.

File: samples/dnn/alpha_matting.cpp (hunk 12 above)
Comment (HarxSan) [reply samples/dnn/alpha_matting.cpp]: Resolved

File: samples/dnn/alpha_matting.cpp (hunk 13 above)
Comment (HarxSan) [reply samples/dnn/alpha_matting.cpp]: Removed putText as it was just about `quit`.

File: samples/dnn/alpha_matting.cpp (hunk 14 above)
Comment (HarxSan) [reply samples/dnn/alpha_matting.cpp]: Done

**Comment by HarxSan** [comment]
Also removed the support for both video and webcam in both the samples.

File: samples/dnn/alpha_matting.py [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+from common import *
+
+def get_args_parser(func_args):
+    backends = ("default", "openvino", "opencv", "vkcom", "cuda")
+    targets = ("cpu", "opencl", "opencl_fp16", "ncs2_vpu", "hddl_vpu", "vulkan", "cuda", "cuda_fp16")
+
+    parser = argparse.ArgumentParser(add_help=False)
+    parser.add_argument('--zoo', default=os.path.join(os.path.dirname(os.path.abspath(__file__)), 'models.yml'),
+                        help='An optional path to file with preprocessing parameters.')
+    parser.add_argument('--input', help='Path to input image file.', required=True)
```
Comment (abhishek-gola) [inline samples/dnn/alpha_matting.py]: Add a default image _messi5.jpg_ present in _samples/data_. Do it for both the samples, and make sure there is an empty line in the end of file.

Comment (HarxSan) [reply samples/dnn/alpha_matting.py]: Done

File: samples/dnn/alpha_matting.cpp [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+
+    const string target_keys = format(
+        "{ target          | cpu | Choose one of target computation devices: "
+        "cpu: CPU target (by default), "
+        "opencl: OpenCL, "
+        "opencl_fp16: OpenCL fp16 (half-float precision), "
+        "vpu: VPU, "
+        "vulkan: Vulkan, "
+        "cuda: CUDA, "
+        "cuda_fp16: CUDA fp16 (half-float precision) }");
```
Comment (abhishek-gola) [inline samples/dnn/alpha_matting.cpp]: This is the supposed to be global i.e. on top of the file.

Comment (HarxSan) [reply samples/dnn/alpha_matting.cpp]: Moved it to top of the file.

**Comment by abhishek-gola** [comment]
LGTM. Test it properly for every edge case, if works perfectly. Move it to opencv.

**Comment by HarxSan** [comment]
Preparing review...

**Comment by HarxSan** [comment]
## PR Reviewer Guide 🔍

Here are some key observations to aid the review process:

<table>
<tr><td>⏱️&nbsp;<strong>Estimated effort to review</strong>: 3 🔵🔵🔵⚪⚪</td></tr>
<tr><td>🧪&nbsp;<strong>No relevant tests</strong></td></tr>
<tr><td>🔒&nbsp;<strong>No security concerns identified</strong></td></tr>
<tr><td>⚡&nbsp;<strong>Recommended focus areas for review</strong><br><br>

<details><summary><a href='https://github.com/HarxSan/opencv/pull/1/files#diff-589f495d7fc7c446ab254635b52b2fc1302a6f6d0af67eb9d09758233e5bc665R77-R78'><strong>Potential Memory Leak</strong></a>

The `loadModel` function does not check if the `net` object is already initialized before assigning a new value. This could lead to a memory leak if the function is called multiple times.
</summary>

```c++
static void loadModel(const string modelPath, String backend, String target, Net &net, EngineType engine)
{
```

</details>

<details><summary><a href='https://github.com/HarxSan/opencv/pull/1/files#diff-a58d2f38460de7f9e4baed72fdbbc6835e202b97bf6d5897a6d4efadbdc8f5f9R118-R120'><strong>Inconsistent Backend/Target Handling</strong></a>

The backend and target handling in `alpha_matting.py` and `alpha_matting.cpp` is inconsistent. Ensure that both files use the same logic for setting backend and target.
</summary>

```python
h, w = img.shape[:2]
font_scale = max(h, w) / 1000.0
thickness = 1
```

</details>

</td></tr>
</table>

**Comment by HarxSan** [comment]
**[Persistent review](https://github.com/HarxSan/opencv/pull/1#issuecomment-3645100984)** updated to latest commit https://github.com/HarxSan/opencv/commit/bdf01b9488e32d7f48cf1c08c55cc5e514c1a940

**Comment by HarxSan** [comment]
Preparing review...

**Comment by HarxSan** [comment]
## PR Code Suggestions ✨

No code suggestions found for the PR.

**Comment by HarxSan** [comment]
## PR Reviewer Guide 🔍

Here are some key observations to aid the review process:

<table>
<tr><td>⏱️&nbsp;<strong>Estimated effort to review</strong>: 3 🔵🔵🔵⚪⚪</td></tr>
<tr><td>🧪&nbsp;<strong>No relevant tests</strong></td></tr>
<tr><td>🔒&nbsp;<strong>No security concerns identified</strong></td></tr>
<tr><td>⚡&nbsp;<strong>Recommended focus areas for review</strong><br><br>

<details><summary><a href='https://github.com/HarxSan/opencv/pull/1/files#diff-589f495d7fc7c446ab254635b52b2fc1302a6f6d0af67eb9d09758233e5bc665R77-R79'><strong>Potential Memory Leak</strong></a>

The `loadModel` function does not check if `net` is already initialized before calling `readNetFromONNX`. This could lead to memory leaks if the function is called multiple times.
</summary>

```c++
static void loadModel(const string modelPath, String backend, String target, Net &net, EngineType engine)
{
    net = readNetFromONNX(modelPath, engine);
```

</details>

<details><summary><a href='https://github.com/HarxSan/opencv/pull/1/files#diff-a58d2f38460de7f9e4baed72fdbbc6835e202b97bf6d5897a6d4efadbdc8f5f9R111-R114'><strong>Inconsistent Backend/Target Handling</strong></a>

The backend and target settings are handled differently in `alpha_matting.py` and `alpha_matting.cpp`. Ensure consistency across both files.
</summary>

```python
net = cv.dnn.readNetFromONNX(args.model, engine)
net.setPreferableBackend(get_backend_id(args.backend))
net.setPreferableTarget(get_target_id(args.target))
return net
```

</details>

</td></tr>
</table>

**Comment by HarxSan** [comment]
### **Ask**❓
"What is this PR about?"

### **Answer:**
This PR adds a sample demonstrating human portrait alpha matting using the MODNet model with OpenCV's DNN module. The sample includes both a C++ and a Python implementation.

**Comment by HarxSan** [comment]
**[Persistent review](https://github.com/HarxSan/opencv/pull/1#issuecomment-3645100984)** updated to latest commit https://github.com/HarxSan/opencv/commit/bdf01b9488e32d7f48cf1c08c55cc5e514c1a940

**Comment by HarxSan** [comment]
**[Persistent review](https://github.com/HarxSan/opencv/pull/1#issuecomment-3645100984)** updated to latest commit https://github.com/HarxSan/opencv/commit/bdf01b9488e32d7f48cf1c08c55cc5e514c1a940

**Comment by HarxSan** [comment]
**[Persistent review](https://github.com/HarxSan/opencv/pull/1#issuecomment-3673875603)** updated to latest commit https://github.com/HarxSan/opencv/commit/bdf01b9488e32d7f48cf1c08c55cc5e514c1a940

**Comment by HarxSan** [comment]
**[Persistent review](https://github.com/HarxSan/opencv/pull/1#issuecomment-3673875603)** updated to latest commit https://github.com/HarxSan/opencv/commit/bdf01b9488e32d7f48cf1c08c55cc5e514c1a940

**Comment by HarxSan** [comment]
Preparing review...

**Comment by HarxSan** [comment]
/review

## 🔍 Additional Issues Found (6 more)

*Supplemental analysis found these additional concerns:*

🔴 **Memory Safety**
   - *File:* `modules/photo/src/seamless_cloning.cpp` (140)
   - *Description:* Potential null pointer dereference in `seamless_cloning.cpp` when calling `obj.illuminationChange`. The `mask` parameter could be null, leading to undefined behavior.

🟡 **Logic Error**
   - *File:* `modules/videoio/src/cap_aravis.cpp` (172)
   - *Description:* In `cap_aravis.cpp`, the `getCaptureDomain` method returns `cv::CAP_ARAVIS` without checking if the camera is actually open. This could lead to incorrect domain identification if the camera was never opened successfully.

🔴 **Resource Leak**
   - *File:* `modules/videoio/src/cap_aravis.cpp` (117)
   - *Description:* In `cap_aravis.cpp`, the destructor of `CvCaptureCAM_Aravis` calls `close()` which might throw an exception. If an exception occurs during the call to `close()`, the object will still be destroyed, potentially leaving resources unreleased.

🔴 **Thread Safety**
   - *File:* `['modules/photo/src/seamless_cloning.cpp', 'modules/videoio/src/cap_aravis.cpp']` (['140', '117'])
   - *Description:* Both `seamless_cloning.cpp` and `cap_aravis.cpp` use shared data (`src`, `mask`, `blend`) across multiple threads without proper synchronization. This can lead to race conditions and inconsistent state.

🟡 **API Misuse**
   - *File:* `modules/photo/src/seamless_cloning.cpp` (120)
   - *Description:* In `textureFlattening`, the `_dst` parameter is created with `_dst.create(src.size(), src.type())` before being used as input for `obj.textureFlatten`. This is likely a mistake since `_dst` should be an output parameter.

🟡 **Error Handling**
   - *File:* `modules/videoio/src/cap_aravis.cpp` (120)
   - *Description:* In `cap_aravis.cpp`, error handling after `open()` is missing. If `open()` fails, subsequent operations on the `CvCaptureCAM_Aravis` object will operate on invalid state, leading to crashes or unexpected behavior.

*Total issues found: PR-Agent + 6 additional = Enhanced comprehensive review*

**Comment by HarxSan** [comment]
**[Persistent review](https://github.com/HarxSan/opencv/pull/1#issuecomment-3719915616)** updated to latest commit https://github.com/HarxSan/opencv/commit/bdf01b9488e32d7f48cf1c08c55cc5e514c1a940

**Comment by HarxSan** [comment]
**[Persistent review](https://github.com/HarxSan/opencv/pull/1#issuecomment-3719915616)** updated to latest commit https://github.com/HarxSan/opencv/commit/bdf01b9488e32d7f48cf1c08c55cc5e514c1a940

**Comment by HarxSan** [comment]
**[Persistent review](https://github.com/HarxSan/opencv/pull/1#issuecomment-3719915616)** updated to latest commit https://github.com/HarxSan/opencv/commit/bdf01b9488e32d7f48cf1c08c55cc5e514c1a940

**Comment by HarxSan** [comment]
**[Persistent review](https://github.com/HarxSan/opencv/pull/1#issuecomment-3719915616)** updated to latest commit https://github.com/HarxSan/opencv/commit/bdf01b9488e32d7f48cf1c08c55cc5e514c1a940

**Comment by HarxSan** [comment]
**[Persistent review](https://github.com/HarxSan/opencv/pull/1#issuecomment-3719915616)** updated to latest commit https://github.com/HarxSan/opencv/commit/bdf01b9488e32d7f48cf1c08c55cc5e514c1a940

**Comment by HarxSan** [comment]
**[Persistent review](https://github.com/HarxSan/opencv/pull/1#issuecomment-3719915616)** updated to latest commit https://github.com/HarxSan/opencv/commit/bdf01b9488e32d7f48cf1c08c55cc5e514c1a940

**Comment by HarxSan** [comment]
**[Persistent review](https://github.com/HarxSan/opencv/pull/1#issuecomment-3719915616)** updated to latest commit https://github.com/HarxSan/opencv/commit/bdf01b9488e32d7f48cf1c08c55cc5e514c1a940
